csrc/multidevice/pipeline_ir.h

```diff
@@ -0,0 +1,132 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+#include <disjoint_set.h>
+#include <fusion.h>
+#include <ir/base_nodes.h>
+#include <multidevice/device_mesh.h>
+#include <multidevice/multidevice.h>
+
+/*
+  This file implements the IRs composing a Pipeline:
+  - PipelineVals
+  - PipelineStage
+  - PipelineCommunication
+*/
+
+namespace nvfuser {
+
+class PipelineStageDescriptor;
+
+/*
+  A PipelineStage represents a Stage of a Pipeline.
+  It is derived from Expr, and represents the composition
+  of all the exprs of the originalFusion to go from the stage's inputs to the
+  stage's output.
+
+  It is instantiated from a list of I/O and a PipelineStageDescriptor (pointing
+  to the originalFusion's Vals contained in the Stage)
+*/
+class TORCH_CUDA_CU_API PipelineStage : public Expr {
+ public:
+  using Expr::Expr;
+
+  PipelineStage(
+      IrBuilderPasskey passkey,
+      const PipelineStageDescriptor* descriptor,
+      ValSet input_vals,
+      ValSet output_vals);
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  std::string toString(int indent_size = 0) const override;
+  std::string toInlineString(int indent_size = 0) const override;
+
+  const char* getOpString() const override {
+    return "PipelineStage";
+  }
+
+  bool sameAs(const Statement* other) const override;
+
+  const PipelineStageDescriptor* descriptor() const {
+    return descriptor_;
+  }
+
+ private:
+  const PipelineStageDescriptor* descriptor_;
```

Comment:
Exprs should not have private members. They should use attributes to store data. For example:
https://github.com/NVIDIA/Fuser/blob/d7d30e680f7916c97ba84300f038907706edbeb7/csrc/ir/nodes.cpp#L151-L152
https://github.com/NVIDIA/Fuser/blob/d7d30e680f7916c97ba84300f038907706edbeb7/csrc/ir/internal_nodes.h#L84-L86